File: variants/haskell/src/RCA/Thread.hs
-------------------------------------------------------------------------------
-- Regulated Cell Architecture (RCA) - Haskell Variant
-- Thread
-------------------------------------------------------------------------------

module RCA.Thread where

import RCA.Data
import RCA.Cell

-------------------------------------------------------------------------------
-- (1) Threads
-------------------------------------------------------------------------------

-- Status: MUTABLE
numThreads :: Int
numThreads = 1

-- Status: MUTABLE
executionThreshold :: Double
executionThreshold = 1.0  -- Units in ms

-- Status: MUTABLE
data Effect = Effect
  { efxActivity :: ActivityInfo
  , efxHandoff  :: CellData
  , efxFinished :: Bool
  } deriving (Show)

-- Status: MUTABLE
data ProgramThread = ProgramThread
  { ptCounter :: Int
  , ptTasks   :: [Cell]
  , ptHandoff :: CellData
  } deriving (Show)

taskName :: Task -> String
taskName TaskDefault     = "Default"
taskName TaskDoubleValue = "DoubleValue"

buildTasks :: Maybe Int -> Maybe [Cell] -> Maybe CellData -> ProgramThread
buildTasks mctr mtsks mho = ProgramThread
  { ptCounter = maybe 0 id mctr
  , ptTasks   = maybe cellDefaults id mtsks
  , ptHandoff = maybe CellNone id mho
  }

isFinished :: ProgramThread -> Bool
isFinished pt = ptCounter pt >= length (ptTasks pt)

step :: ProgramThread -> DataPlane -> (ProgramThread, Effect)
step pt ctx =
  let currentCell = ptTasks pt !! ptCounter pt
      activity    = ActivityInfo { actDescription = taskName (cellTask currentCell) }
      -- Handoff transfer: take current, pass to cell, store result back
      newHandoff  = cellExecute currentCell ctx (ptHandoff pt)
      pt'         = pt { ptCounter = ptCounter pt + 1
                       , ptHandoff = newHandoff
                       }
      finished    = isFinished pt'
      effect      = Effect
                       { efxActivity = activity
                       , efxHandoff  = newHandoff
                       , efxFinished = finished
                       }
  in (pt', effect)

File: variants/javascript/rca/thread.js
///////////////////////////////////////////////////////////////////////////////
// Regulated Cell Architecture (RCA) - JavaScript Variant
// Thread
///////////////////////////////////////////////////////////////////////////////

const { createActivityInfo } = require("./data");
const { NUM_CELLS, cellNone, cellExecute, cellDefaults } = require("./cell");

///////////////////////////////////////////////////////////////////////////////
// (1) Threads
///////////////////////////////////////////////////////////////////////////////

// Status: MUTABLE
const NUM_THREADS = 1;

// Status: MUTABLE
const EXECUTION_THRESHOLD = 1.0; // Units in ms

// Status: MUTABLE
function createProgramThread(tasks, handoff) {
  return {
    counter: 0,
    tasks:   tasks || cellDefaults(),
    handoff: handoff || cellNone(),
  };
}

function buildTasks(tasks, handoff) {
  return createProgramThread(tasks, handoff);
}

function isFinished(pt) {
  return pt.counter >= pt.tasks.length;
}

function step(pt, ctx) {
  const activity = createActivityInfo(pt.tasks[pt.counter].task);

  // Handoff transfer: take current, pass to cell, store result back
  const handoffTransfer = pt.handoff;
  pt.handoff = cellNone();

  pt.handoff = cellExecute(pt.tasks[pt.counter], ctx, handoffTransfer);
  pt.counter++;

  const finished = isFinished(pt);

  return {
    activity,
    handoff: pt.handoff,
    finished,
  };
}

module.exports = {
  NUM_THREADS,
  EXECUTION_THRESHOLD,
  createProgramThread,
  buildTasks,
  isFinished,
  step,
};
